### PR TITLE
Avoid returning checksums unnecessarily during OOP calls.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Our existing model is one where we send over a bunch of checksums, then send those checksums back with the objects they are attached to.  None of this is necessary as we know exactly what checksums we're sending over, and thus know exactly which ones the responses are associated with.

### CODE REVIEW

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs [hunk 1]
```diff
@@ -38,60 +38,57 @@ internal static class RemoteHostAssetSerialization
...
-                WriteAsset(writer, serializer, context, checksums[0], singleAsset, cancellationToken);
+                WriteAsset(writer, serializer, context, singleAsset, cancellationToken);
                 return;
             }
 
             Debug.Assert(assetMap != null);
-            writer.WriteInt32(assetMap.Count);
 
-            foreach (var (checksum, asset) in assetMap)
+            var count = 0;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:51]:
```suggestion
```

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs [hunk 2]
```diff
@@ -38,60 +38,57 @@ internal static class RemoteHostAssetSerialization
...
+            foreach (var checksum in checksums)
             {
-                WriteAsset(writer, serializer, context, checksum, asset, cancellationToken);
+                var asset = assetMap[checksum];
+
+                WriteAsset(writer, serializer, context, asset, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:56]:
```suggestion                // We flush after each item as that forms a reasonably sized chunk of data to want to then send over the
                // pipe for the reader on the other side to read.  This allows the item-writing to remain entirely
                // synchronous without any blocking on async flushing, while also ensuring that we're not buffering the
                // entire stream of data into the pipe before it gets sent to the other side.
                WriteAsset(writer, serializer, context, asset, cancellationToken);
```

File: src/Workspaces/CoreTestUtilities/Fakes/SimpleAssetSource.cs [hunk 3]
```diff
@@ -10,52 +10,42 @@
...
     {
-        private readonly ISerializerService _serializerService;
-        private readonly IReadOnlyDictionary<Checksum, object> _map;
-
-        public SimpleAssetSource(ISerializerService serializerService, IReadOnlyDictionary<Checksum, object> map)
-        {
-            _serializerService = serializerService;
-            _map = map;
-        }
+        var results = new List<object>();
```
Comment (ToddGrun) [inline src/Workspaces/CoreTestUtilities/Fakes/SimpleAssetSource.cs:23]: >new List<object>(); [](http://example.com/codeflow?start=22&length=19)

nit: Initialize with size or consider using pool?

Comment (CyrusNajmabadi) [reply src/Workspaces/CoreTestUtilities/Fakes/SimpleAssetSource.cs:23]: test code :)

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs [hunk 4]
```diff
@@ -38,60 +38,53 @@ internal static class RemoteHostAssetSerialization
...
-                {
                     serializer.Serialize(asset.Value, writer, context, cancellationToken);
-                }
             }
         }
 
-        public static async ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(
-            PipeReader pipeReader, Checksum solutionChecksum, ISet<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
+        public static async ValueTask<ImmutableArray<object>> ReadDataAsync(
+            PipeReader pipeReader, Checksum solutionChecksum, ImmutableArray<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
```
Comment (ToddGrun) [inline src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:77]: >ImmutableArray<Checksum> checksums [](http://example.com/codeflow?start=62&length=34)

This just needs a length, not the actual array anymore, right?

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs]: nice catch!

**Pull request merged**